Analysis Of Hacker News

Hacker News is a site started by the startup,where user-submitted stories (known as "posts") receive votes and comments, similar to reddit.We're specifically interested in posts with titles that begin with either Ask HN or Show HN. 

We'll compare these two types of posts to determine the following:
1)Do Ask HN or Show HN receive more comments on average?

2)Do posts created at a certain time receive more comments on average?


In [1]:
from csv import reader
with open("hacker_news.csv") as file:
    read_file=reader(file)
    hn=list(read_file)
five_rows=hn[:5]
print(five_rows)

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


In [2]:
headers=hn[:1]
print(headers)


[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']]


In [3]:
del hn[:1]

In [4]:
print(hn[:5])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


In [5]:
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1].lower()
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else:
        other_posts.append(row)
    

In [6]:
print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))

1744
1162
17194


In [7]:
total_ask_comments=0
for data in ask_posts:
    num_comments=data[4]
    int_num_comments=int(num_comments)
    total_ask_comments+=int_num_comments
avg_ask_comments=total_ask_comments/len(ask_posts)
print(avg_ask_comments)

14.038417431192661


In [8]:
total_show_comments=0
for data in show_posts:
    num_comments=data[4]
    int_num_comments=int(num_comments)
    total_show_comments+=int_num_comments
avg_show_comments=total_show_comments/len(show_posts)
print(avg_show_comments)

10.31669535283993


Ask posts get a more commments than a show posts on average. so high number of peopless are interested into Ask posts to solve the problem of people rather than appreciating and aknowledging the work 



In [9]:
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = row[4]
    int_num_comments = int(num_comments)  # fixed variable name
    result_list.append((created_at, int_num_comments))  # fixed variable name

counts_by_hour = {}
comments_by_hour = {}

for data in result_list:
    dt_object = dt.datetime.strptime(data[0], "%m/%d/%Y %H:%M")
    hour = dt_object.hour

    if hour not in counts_by_hour:
        counts_by_hour[hour] = 1
        comments_by_hour[hour] = data[1]  # Use actual comment count here
    else:
        counts_by_hour[hour] += 1
        comments_by_hour[hour] += data[1]  # Add the actual comment count


In [10]:
print(counts_by_hour)
print(comments_by_hour)

{9: 45, 13: 85, 10: 59, 14: 107, 16: 108, 23: 68, 12: 73, 17: 100, 15: 116, 21: 109, 20: 80, 2: 58, 18: 109, 3: 54, 5: 46, 19: 110, 1: 60, 22: 71, 8: 48, 4: 47, 0: 55, 6: 44, 7: 34, 11: 58}
{9: 251, 13: 1253, 10: 793, 14: 1416, 16: 1814, 23: 543, 12: 687, 17: 1146, 15: 4477, 21: 1745, 20: 1722, 2: 1381, 18: 1439, 3: 421, 5: 464, 19: 1188, 1: 683, 22: 479, 8: 492, 4: 337, 0: 447, 6: 397, 7: 267, 11: 641}


In [15]:
avg_by_hour = []

for hour in counts_by_hour:
    avg = comments_by_hour[hour] / counts_by_hour[hour]
    avg_by_hour.append([hour, avg])


In [16]:
print(avg_by_hour[:5])

[[9, 5.5777777777777775], [13, 14.741176470588234], [10, 13.440677966101696], [14, 13.233644859813085], [16, 16.796296296296298]]


In [21]:
swap_avg_by_hour=[]
for row in avg_by_hour:
    
    swap_avg_by_hour.append([row[1],row[0]])

In [22]:
print(swap_avg_by_hour)

[[5.5777777777777775, 9], [14.741176470588234, 13], [13.440677966101696, 10], [13.233644859813085, 14], [16.796296296296298, 16], [7.985294117647059, 23], [9.41095890410959, 12], [11.46, 17], [38.5948275862069, 15], [16.009174311926607, 21], [21.525, 20], [23.810344827586206, 2], [13.20183486238532, 18], [7.796296296296297, 3], [10.08695652173913, 5], [10.8, 19], [11.383333333333333, 1], [6.746478873239437, 22], [10.25, 8], [7.170212765957447, 4], [8.127272727272727, 0], [9.022727272727273, 6], [7.852941176470588, 7], [11.051724137931034, 11]]


In [23]:
sorted_swap=sorted(swap_avg_by_hour,reverse=True)


In [24]:
print(sorted_swap[:5])

[[38.5948275862069, 15], [23.810344827586206, 2], [21.525, 20], [16.796296296296298, 16], [16.009174311926607, 21]]


In [25]:
for avg, hour in sorted_swap[:5]:
        time_obj = dt.datetime.strptime(str(hour), "%H")
        formatted_time = time_obj.strftime("%H:00")
        print(f"{formatted_time}: {avg:.2f} average comments per post")

15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set documentation, the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.

Conclusion
In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.